datacube_alchemist/worker.py

```diff
@@ -231,6 +213,101 @@ def post_messages(messages, count):
+        # Retrieve a task based on a UUID, or none if it doesn't exist for input product(s)
+        dataset = self._find_dataset(uuid)
+        if dataset:
+            return AlchemistTask(dataset=dataset, settings=self.config)
+        else:
+            return None
+
+    def generate_tasks(self, query, limit=None) -> Iterable[AlchemistTask]:
+        # Find which datasets needs to be processed
+        datasets = self._find_datasets(query, limit)
+
+        tasks = (self.generate_task(ds) for ds in datasets)
+
+        return tasks
+
+    # Queue related functions
+    def enqueue_datasets(self, queue, query, limit=None, product_limit=None):
+        datasets = self._find_datasets(query, limit, product_limit)
+        return self._datasets_to_queue(queue, datasets)
+
+    def find_fill_missing(self, queue, dryrun):
+        query = """
+            select d.id
+            from agdc.dataset as d
+            join agdc.dataset_type as t
+            on d.dataset_type_ref = t.id
+            left join (
+                select
+                    sub_s.dataset_ref,
+                    sub_s.source_dataset_ref
+                from agdc.dataset_source as sub_s
+                join agdc.dataset as sub_d
+                on sub_d.id = sub_s.dataset_ref
+                join agdc.dataset_type as sub_t
+                on sub_d.dataset_type_ref = sub_t.id
+                where sub_t.name = '{output_product}'
+            ) as s
+            on d.id = s.source_dataset_ref
+            where t.name in ({input_products})
+            and s.dataset_ref is null
+        """
+
+        # Most of this guff is just to get a destination product name...
+        input_products = ", ".join(f"'{p.name}'" for p in self.input_products)
+        output_product = ""
+
+        dataset = self.dc.find_datasets(product=self.input_products[0].name, limit=1)
+        source_doc = _munge_dataset_to_eo3(dataset[0])
+
+        with DatasetAssembler(
+            metadata_path="/fake/path",
+            naming_conventions=self.naming_convention,
+        ) as dataset_assembler:
+            dataset_assembler.add_source_dataset(
+                source_doc,
+                auto_inherit_properties=True,
+                inherit_geometry=self.config.output.inherit_geometry,
+                classifier=self.config.specification.override_product_family,
+            )
+            for k, v in self.config.output.metadata.items():
+                setattr(dataset_assembler, k, v)
+            for k, v in self.config.output.properties.items():
+                dataset_assembler.properties[k] = v
+            dataset_assembler.processed = datetime.utcnow()
+
+            output_product = dataset_assembler.names.product_name
+            dataset_assembler.cancel()
+
+        # This is actual valuable work
+        _LOG.info("Querying database, please wait...")
+        conn = psycopg2.connect(str(self.dc.index.url))
```

Comment:
I think I prefer `c = dc.index._db.give_me_a_connection()` and using the SQLAlchemy connection like `results = list(c.execute(query, input_products=input_products, output_product=output_product))`.

But, it's not a strong preference at all, so nevermind. :)

Reply:
Agreed. A downside of this approach is that it's constantly connecting and reconnecting to the db, while the datacube's sqlalchemy already has a connection (& pool) ready to reuse the same connection(s). This may or may not be significant depending on the environment being deployed to.

Reply:
It also loses the "_extremely important_" application_name in the connection parameters :slightly_smiling_face:

Reply:
This command is run very rarely and I think keeping it like this is fine.